streaming/cleanup.go

```diff
@@ -0,0 +1,221 @@
+package streaming
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+
+	dataflow "cloud.google.com/go/dataflow/apiv1beta3"
+	datastream "cloud.google.com/go/datastream/apiv1"
+	"cloud.google.com/go/pubsub"
+	"cloud.google.com/go/spanner"
+	"cloud.google.com/go/storage"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/common/utils"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/internal"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/logger"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/profiles"
+	"google.golang.org/api/iterator"
+)
+
+type ShardCleanupOptions struct {
+	Dataflow   bool
+	Datastream bool
+	Pubsub     bool
+	Monitoring bool
+}
+
+type JobCleanupOptions struct {
+	Monitoring bool
+}
+
+func InitiateJobCleanup(ctx context.Context, options JobCleanupOptions, jobExecutionData JobExecutionData, project string) error {
+	if options.Monitoring {
+		logger.Log.Debug("Attempting to delete aggregate monitoring resources...\n")
+		var aggMonitoringResources internal.MonitoringResources
+		err := json.Unmarshal([]byte(jobExecutionData.AggMonitoringResources), &aggMonitoringResources)
+		if err != nil {
+			logger.Log.Debug("Unable to read aggregate monitoring metadata for deletion\n")
+			return err
+		}
+		CleanupMonitoringDashboard(ctx, aggMonitoringResources.DashboardName, project)
+	}
+	return nil
+}
+
+func InitiateShardCleanup(ctx context.Context, options ShardCleanupOptions, shardExecutionData ShardExecutionData, project string, instance string) error {
+	if options.Dataflow {
+		logger.Log.Debug("Attempting to delete dataflow job...\n")
+		c, err := dataflow.NewJobsV1Beta3Client(ctx)
+		if err != nil {
+			return fmt.Errorf("could not create job client: %v", err)
+		}
+		defer c.Close()
+		logger.Log.Debug("Created dataflow job client...")
+		var dataflowResources internal.DataflowResources
+		err = json.Unmarshal([]byte(shardExecutionData.DataflowResources), &dataflowResources)
+		if err != nil {
+			logger.Log.Debug("Unable to read Dataflow metadata for deletion\n")
+			return err
+		}
+		err = CleanupDataflowJob(ctx, c, dataflowResources.JobId, project, dataflowResources.Region)
+		if err != nil {
+			logger.Log.Debug(fmt.Sprintf("Cleanup of the dataflow job: %s was unsuccessful, please clean up the job manually. err = %v\n", dataflowResources.JobId, err))
+		} else {
+			logger.Log.Info("Dataflow job cleaned up successfully.\n")
+		}
+	}
+	if options.Datastream {
+		logger.Log.Debug("Attempting to delete datastream stream...\n")
+		dsClient, err := datastream.NewClient(ctx)
```

Comment:
Wouldn't this client be created everytime? There is a common getDatastreamClient method in conversion.go that can be used

Reply:
Yeah it would, I just followed the existing approach in HB of creating client on demand.

Reply:
I had to use a bunch of other clients for cleanup (as done in other options) so I thought might as well keep it uniform for now and then refactor later

Reply:
Okay sure